internal/template-validator/tlsinfo/tlsinfo_test.go

```diff
@@ -15,79 +16,196 @@ import (
+
 var _ = Describe("TlsInfo", func() {
-	var certDir string
+	var (
+		certDir       string
+		tlsOptionsDir string
+
+		tlsOptions *common.SSPTLSOptions
+
+		expectedVersion      uint16
+		expectedCipherSuites []uint16
+
+		tlsInfo *TLSInfo
+	)
 
 	BeforeEach(func() {
+		certDir = GinkgoT().TempDir()
+		tlsOptionsDir = GinkgoT().TempDir()
+
+		writeCertificate(certDir)
+
 		var err error
-		certDir, err = os.MkdirTemp("", "certs")
+		tlsOptions, err = common.NewSSPTLSOptions(
+			&ocpconfigv1.TLSSecurityProfile{
+				Type:         ocpconfigv1.TLSProfileIntermediateType,
+				Intermediate: &ocpconfigv1.IntermediateTLSProfile{},
+			},
+			nil,
+		)
 		Expect(err).ToNot(HaveOccurred())
-	})
 
-	It("should return nil if no certificate exists", func() {
-		tlsInfo := TLSInfo{CertsDirectory: certDir}
-		tlsInfo.InitCerts()
-		defer tlsInfo.Clean()
-		tlsConfig := tlsInfo.CreateTlsConfig()
+		writeTLSOptions(tlsOptionsDir, tlsOptions)
 
-		Consistently(func() *tls.Certificate {
-			cert, _ := tlsConfig.GetCertificate(nil)
-			return cert
-		}, time.Second).Should(BeNil())
-	})
+		expectedVersion = tls.VersionTLS12
+		expectedCipherSuites = []uint16{
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+			tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
+			tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
+		}
 
-	It("should load certificate", func() {
-		writeCertificate(certDir)
-		tlsInfo := TLSInfo{CertsDirectory: certDir}
-		tlsInfo.InitCerts()
-		defer tlsInfo.Clean()
-		tlsConfig := tlsInfo.CreateTlsConfig()
-
-		Eventually(func() (*tls.Certificate, error) {
-			return tlsConfig.GetCertificate(nil)
-		}, time.Second).ShouldNot(BeNil())
+		tlsInfo = &TLSInfo{
+			CertsDirectory:      certDir,
+			TLSOptionsDirectory: tlsOptionsDir,
+		}
 	})
 
-	It("should reload new certificate", func() {
-		tlsInfo := TLSInfo{CertsDirectory: certDir}
-		tlsInfo.InitCerts()
-		defer tlsInfo.Clean()
-		tlsConfig := tlsInfo.CreateTlsConfig()
+	Context("loading certificates", func() {
+		It("should fail if no certificate exists", func() {
+			Expect(os.Remove(filepath.Join(certDir, CertFilename))).To(Succeed())
+			Expect(os.Remove(filepath.Join(certDir, KeyFilename))).To(Succeed())
 
-		Consistently(func() *tls.Certificate {
-			cert, _ := tlsConfig.GetCertificate(nil)
-			return cert
-		}, time.Second).Should(BeNil())
+			Expect(tlsInfo.Init()).To(Succeed())
+			defer tlsInfo.Clean()
 
-		writeCertificate(certDir)
+			Consistently(func() error {
+				_, err := tlsInfo.CreateTlsConfig()
+				return err
+			}, time.Second).Should(HaveOccurred())
+		})
 
-		Eventually(func() (*tls.Certificate, error) {
-			return tlsConfig.GetCertificate(nil)
-		}, time.Second).ShouldNot(BeNil())
-	})
+		It("should load certificate", func() {
+			Expect(tlsInfo.Init()).To(Succeed())
+			defer tlsInfo.Clean()
```

Comment:
`defer` vs `DeferCleanup`?

Reply:
`defer` is called when `func() {}` returns. `DeferCleanup()` is called after the `It()` test case, similarly to `AfterEach()`.
 It is useful to call in `BeforeEach()` to register cleanups that will be called after the test.

Reply:
Does it make a difference using `DeferCleanup` here?

Reply:
I think only a performance difference.

Reply:
Can it be moved to `AfterEach()`?

Reply:
Yes, but I prefer it this way. It is also how it's used in the validator code `app.go`.